Addressing review comments

kahaaga opened the pull request:
Hey, @Datseris. I've tried to address all your comments on #96  in this PR. 

- Added `alphabet_length` function that dispatches on estimators (throws error if alphabet length cannot be obtained)
- Standardised API for normalized Tsallis, Renyi and Shannon entropies.
- Added literature references justifying that normalization for symbolic entropy estimators are valid for all orders `q`.
- Made `distance_to_whitenoise` part of public API. 
- Some minor formatting and language fixes.

Some things to potentially discuss are:

## Naming (normalization interface)

- Instead of a single `entropy_normalized` function, I used `entropy_renyi_norm`,
    `entropy_tsallis_norm` and `entropy_shannon_norm`. This mirrors the `entropy_renyi` and 
    similar functions we already have. The alternative is to have `entropy_normalized(x, est = Renyi(); kwargs...)` or something along those lines. As we already discussed elsewhere, I don't think the latter approach is a good idea.
- Do we provide normalized version of the convenience functions too (`entropy_permutation_norm`), or leave it up to the user to figure out how to do so?  

## Normalization/alphabet length

Where to indicate which estimators implements `alphabet_length`? Alternatives are:

- To throw an error in `alphabet_length` for estimators that don't support normalization, e.g. `throw("unimplemented")`. I favor explicitly throwing an informative error instead of silently 
returning `nothing` when an estimator doesn't implement `alphabet_length`, because this 
reduces changes of errors for users.
- In the docstring for `renyi_entropy_norm`.
- In each individual estimator under a `##Implements` heading.
- A combination.

For user-friendlyness, I think either of the latter is best. For code maintenance purposes, the first alternative is best (which is what I've done in this PR).

Datseris commented on src/probabilities_estimators/utils.jl:
````diff
@@ -0,0 +1,27 @@
+## Examples
+
+```jldoctest setup = :(using Entropies)
+julia> est = SymbolicPermutation(m = 4)
+SymbolicPermutation{typeof(Entropies.isless_rand)}(1, 4, Entropies.isless_rand)
````
Put `;` here at the end of `est` assignment?

Datseris commented on src/probabilities_estimators/utils.jl:
```diff
@@ -0,0 +1,27 @@
+
+alphabet_length(est) =
+    throw(error("alphabet_length not implemented for estimator of type $(typeof(est))"))
+alphabet_length(est::SymbolicPermutation)::Int = factorial(est.m)
+alphabet_length(est::SymbolicWeightedPermutation)::Int = factorial(est.m)
+alphabet_length(est::SymbolicAmplitudeAwarePermutation)::Int = factorial(est.m)
```
This function shoould be extended in the same file the `probabilities` function is extended, i.e., in the file of the estimator.

Datseris commented on src/probabilities_estimators/utils.jl:
````diff
@@ -0,0 +1,27 @@
+24
+```
+"""
+function alphabet_length end
+
+alphabet_length(est) =
````
These two definitions should be combined. Also the second definition should type annotate `est::ProbabilitiesEstimator`.

Datseris reviewed (commented):
> Naming (normalization interface)

We don't have to do it this way, we can still have a "single function" without introducing any new stuff like `Renyi`. The key is that in Julia functions are their own types. Thus, we can have a single function
```julia
function normalize(entropy_function, args...; kwargs...)
end
```
which returns the normalized entropy. Instead of the common name `normalize` we could have `entropy_normalized` instead. You define method dispatch as such:
```
function entropy_normalized(entropy_function::typeof(entropy_renyi), args...; kwargs...)
end
```
and this is a new method. 

There are some advantages in this approach:

1. We define a single API function with a single and rather consise docstring. It practically says that the entropy is normalized versus the entropy of uniform distribution across alphabet.
2. If we add more entrpy methods in the future, we can add them to this existing function instead.
3. We can have a cleaner code by defining this one function in its separate file.

I am thinking that we need to define _only a single function_. Given the alphabet length, we can generate a `Probabilities` vector with all entries equal, and pass that to the entropy function directly, no matter what this entropy function is!

> (throws error if alphabet length cannot be obtained)

Actually in the docs you say it returns `nothing` already. So the question is, what should it do by default? At the moment it errors by default, which I'm still not sure. I guess it's okay to leave it as is.

Datseris commented:
Here's my idea for a generic function:
```julia
function normalize(entropy_function, est::ProbabilitiesEstimator, args...; kwargs...)
    L = alphabet_length(est)
    np = Probabilities(ones(L))
    return entropy_function(est, args...; kwargs...)/entropy_function(np, args...; kwargs...)
end
```

kahaaga commented:
> Here's my idea for a generic function:

Excellent. I didn't think of that at all. I will adjust the PR accordingly and ping you when ready.

Datseris commented:
The citation 
```
# Normalization is well-defined for all values of `q`, e.g. Kumar, U., Kumar, V., &
# Kapur, J. N. (1986). Normalized measures of entropy. International Journal Of General
# System, 12(1), 55-69.
```
should be included in the docstring of this function I guess.

kahaaga commented:
Thinking a bit more about this, I'm not convinced that your pseudocode above is the best way to go.  There are two reasons for this: 
- We implicitly assume that *all* generalized entropies are maximized for a uniform distribution. I don't know of any proof that this is the case. For example, is the Kaniadakis entropy (#92) maximized for the uniform distribution?
- If a generalized entropy `E` shows up that is maximized for some other distribution, do we introduce a new normalization function specifically for `E` (e.g. `normalized_X`)?
- For entropies that actually are maximized for uniform distributions (the ones we currently have), why take the extra step of allocating an extra probability vector and computing the entropy of those probabilities too? This essentially doubles the computational cost of computing normalized entropies, when it can be done with a single division operation (dividing by `log(alphabet_length(est))`), like so:

```julia
function normalize(x, entropy_function, est::ProbabilitiesEstimator, args...; kwargs...)
    L = alphabet_length(est)
    np = Probabilities(ones(L))
    return entropy_function(x, est, args...; kwargs...)/log(L)
end
```

Datseris commented:
> We implicitly assume that all generalized entropies are maximized for a uniform distribution

No, we never assume this. We return the entropy normalized versus the uniform distribution. We don't make claims of what the uniform distribution does.

> If a generalized entropy E shows up that is maximized for some other distribution, do we introduce a new normalization function specifically for E (e.g. normalized_X)?

No, because maximizing was never the goal here. Normalizing versus an established standard is.

> This essentially doubles the computational cost of computing normalized entropies

The cost of computing the entropy includes computing the probabilities. Here we just create a small vector, which I bet would have minimal cost versus that of actually computing the probabilities... But sure, if you are analytically certain that for all entropy methods, regardless of `q`, we get `log(alphabet_length(est))` as the result, the use that, definitely. I don't know this for sure, and I don't want to study all these Tsallis entropies now to find out :D

kahaaga commented:
> No, we never assume this. We return the entropy normalized versus the uniform distribution. We don't make claims of what the uniform distribution does.

Sure, but the whole point of normalisation is to ensure that the computed value falls in the range `[0, 1]`, so that entropy values can be compared. This happens precisely when normalising to the maximal value of the entropy. For the entropies we have now, normalising to the uniform distribution *does* normalize to `[0, 1]`, but only because the uniform distribution is the one that maximizes Renyi and Tsallis entropies (according to the references in this PR).

For other entropies, this might not be the case. If there is suddenly an entropy that isn't maximised by the uniform distribution, then the normalized value is no longer guaranteed to be on `[0, 1]`, but on `[0, a]`, where `a` is determined by how that particular entropy behaves for the uniform distribution. Then `entropy_normalized` no longer behaves in a consistent manner. 

Of course, we can just define normalization in Entropies.jl to *mean* normalization to the entropy of a uniform distribution (as you suggest), and leave it to the user to find out if this normalization is meaningful to them . However, I think we should strive to provide "meaningful" output out-of-the-box, which makes having a completely generic `entropy_normalized` function potentially problematic.

> But sure, if you are analytically certain that for all entropy methods, regardless of q, we get `log(alphabet_length(est))` as the result, the use that, definitely.

Based on the references I found, this is the case for Renyi and Tsallis entropies. Not sure about the Kaniadakis and entropies, though. But this is a potential issue regardless if we normalize to `log(alphabet_length(est))` or manually construct a flat distribution and normalize to the entropy of that.

Datseris commented:
> What do you think?

That we are spending way too much time thinking about this :P 

Let's just say, clearly, in the docstring that we divide by `log(alphabet_length`, and that this corresponds to maximum value for the two entropies we already have, and be done with it.

Datseris commented:
new entropies will get same treatment, same division, and the user needs to be educated about whether this makes sense for them I think...

kahaaga commented:
> That we are spending way too much time thinking about this :P

Fair enough 😆 

> Let's just say, clearly, in the docstring that we divide by log(alphabet_length, and that this corresponds to maximum value for the two entropies we already have, and be done with it.

Ok, done. I'll tag you once I've implemented the new interface.

kahaaga commented:
Changes since the last comments:
- `alphabet_length(est)` throws error when the alphabet length is not defined for `est`.
- Moved `alphabet_length` functions to the individual estimator files.
- Added analytical tests for Tsallis entropy, based on the original paper. It turned out Tsallis himself had a more general normalization that I use instead (works for all `q`). Then, dividing by `log(alphabet_length(est))` would no longer work, so I chose to divide by the entropy of a manually constructed flat distribution instead, like you initially proposed. It is still possible to dispatch on particular estimators if we need to reduce the computational cost of the extra allocation and entropy computation.
 - Normalization interface as discussed above, with signatures covering the use cases I imagine I'd want as a user.
- Utility functions `maxentropy_tsallis`, `maxentropy_renyi`, `maxentropy_shannon` (listed under "Utility methods" in the docs) that can be used for manual normalization.

Datseris commented on src/entropies/normalized_entropy.jl:
```diff
@@ -0,0 +1,80 @@
+Normalization is only defined for estimators for which [`alphabet_length`](@ref) is defined,
+meaning that the total number of states or symbols is known beforehand.
+
+    entropy_normalized(f::Function, p::Probabilities, est::ProbabilitiesEstimator, args...;
+        kwargs...)
+
+Normalize the entropy, as returned by the entropy function `f` called with the given
+arguments (i.e. `f(p, args...; kwargs...)`), to the entropy of a uniform distribution,
+inferring [`alphabet_length`](@ref) from `est`.
```
I'm not convinced that this method is useful.... If the user has the `est` they may as well use the first method.

kahaaga replied:
We can just remove it. The user can just normalize to `log(N)` manually if they need to.

kahaaga replied:
Wait, I'm not so sure of that. If we remove this method, then:

Say I have a set of probabilities `p = Probabilities([0.2, 0.2, 0.3, 0.3])` *that I have already computed with a known estimator, but I don't have the original data for*. I now want to compute 1) Renyi entropy and 2) normalized Renyi entropy. Without this method, it is easy to compute 1), but no convenient way to compute 2), without re-implementing precisely this function does myself. That is: I'd always have to start from raw data if wanting to compute normalized entropy.

Datseris replied:
What do you mean, you do `entropy_renyi(p)/log(length(p))`. What's the problem :D

kahaaga replied:
> What do you mean, you do entropy_renyi(p)/log(length(p)). What's the problem :D

[See this comment](https://github.com/JuliaDynamics/Entropies.jl/pull/99#discussion_r980261798).

Datseris commented on src/entropies/renyi.jl:
```diff
@@ -89,3 +94,14 @@ function entropy_renyi!(p, x, est; q = 1.0, α = nothing, base = MathConstants.e
+
+See also [`entropy_renyi`](@ref), [`entropy_normalized`](@ref).
+"""
+maxentropy_renyi(N::Int; base = MathConstants.e) = log(base, N)
```
Why would we need such a function? I'm not convinced that providing a wrapper function over `log` is useful either :D

kahaaga replied:
Because maximum entropy in itself is a useful thing to be able to compute. For Renyi entropy, it is trivially `log(N)`, but for other entropies, it is not trivial at all. For stretched exponential entropy, for example, which I've prepared in a non-committed PR, the maximum entropy is a function of the parameters of the method and involves complicated expression involving gamma and incomplete gamma functions. 

I think there should be a convenient way for the user to compute this quantity, to also be able to work with the entropies from a theoretical standpoint. All analytical tests for this stretched exponential entropy, and the example reproducing the original paper, for example, involve using `maxentropy_stretched_exponential`. 

The maximum entropy is a fundamental property of entropies. I don't see why we shouldn't provide this functionality. This is after all a package that is dedicated specifically to computing entropies and related quantities.  I don't think pruning is necessary just because we don't see an immediate use for the method at the moment.

Datseris replied:
If you also vote for that I'd like to do it after I merge this PR.

kahaaga replied:
> Actually, this whole business makes me think if we should make "entropy types". We can have a single entropy(...) function, and we have "entropy types" such as Renyi() and Tsallis(). So the renyi entropy for a given estimator would be entropy(Renyi(), x, est) This way we can actually call maximum(Renyi(), PermutationX()) instead of passing functions. The estimators can have arguments such as Renyi(q) for the order q.

Yes, going for the dispatch version would perhaps be better once the codebase grows. I wasn't aware to begin with how many types of entropies there actually are, but they are in fact plentiful.

If the rest of this PR looks ok to you, maybe we just merge this PR (in the current state) and the original one to `main`.  Then, as you say, we can implement the multiple dispatch `entropy(EntropyType(), Estimator())` version in a separate PR. 

If this is merged quickly, I can put together a PR tonight for the new version, because I'd like to get this off the table as soon as possible and arrive at a stable API.

Datseris commented:
I'd still cut down on content here. E.g, i don't think providing `maxentropy_renyi`, which literally is the `log`, makes sense. Let me know if you want me to do the pruning here or if yo uwant to continue working here instead.

kahaaga commented:
> I'd still cut down on content here. E.g, i don't think providing maxentropy_renyi, which literally is the log, makes sense. we can mention it in the docstring isteadn. Let me know if you want me to do the pruning here or if yo uwant to continue working here instead.

See comment above

Datseris commented:
Let me do some pruning here, and if I prune too much, you let me know and we add back the things you believe are useful.

Datseris commented:
@kahaaga I'd prefer to make the first draft for the new API if that's okay with you and you can review it instead. I'll have it done by tonight.

Datseris commented:
@kahaaga Actually, wait, I'm confused, where was this PR just got merged into? Not at the `api_refactor`, right? Because I was going to update that PR to the new API!

kahaaga commented:
> @kahaaga Actually, wait, I'm confused, where was this PR just got merged into? Not at the api_refactor, right? Because I was going to update that PR to the new API!

This was merged into the `dispersion_estimators` branch. 

> @kahaaga I'd prefer to make the first draft for the new API if that's okay with you and you can review it instead. I'll have it done by tonight.

Perfect.

Datseris commented:
Forget everything I said, I haven't realized that `refactor_api` was already merged into main. We should merge #96 into `main` as well and I start a new pr with the new, new API.

kahaaga commented:
> Forget everything I said, I haven't realized that refactor_api was already merged into main. We should merge https://github.com/JuliaDynamics/Entropies.jl/pull/96 into main as well and I start a new pr with the new, new API.

Yes, precisely.
